**1. Data Inspection**

In [1]:
import pandas as pd

# Load the dataset
df = pd.read_csv("retail_store_sales task 3.csv")

# Display first 5 rows
df.head()

,Transaction ID,Customer ID,Category,Item,Price Per Unit,Quantity,Total Spent,Payment Method,Location,Transaction Date,Discount Applied
0,TXN_6867343,CUST_09,Patisserie,Item_10_PAT,18.5,10.0,185.0,Digital Wallet,Online,2024-04-08,True
1,TXN_3731986,CUST_22,Milk Products,Item_17_MILK,29.0,9.0,261.0,Digital Wallet,Online,2023-07-23,True
2,TXN_9303719,CUST_02,Butchers,Item_12_BUT,21.5,2.0,43.0,Credit Card,Online,2022-10-05,False
3,TXN_9458126,CUST_06,Beverages,Item_16_BEV,27.5,9.0,247.5,Credit Card,Online,2022-05-07,NaN
4,TXN_4575373,CUST_05,Food,Item_6_FOOD,12.5,7.0,87.5,Digital Wallet,Online,2022-10-02,False


In [2]:
# Check dataset size
print("Number of rows:", df.shape[0])
print("Number of columns:", df.shape[1])

Number of rows: 12575
Number of columns: 11


In [3]:
# Check column names and data types
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12575 entries, 0 to 12574
Data columns (total 11 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Transaction ID    12575 non-null  object 
 1   Customer ID       12575 non-null  object 
 2   Category          12575 non-null  object 
 3   Item              11362 non-null  object 
 4   Price Per Unit    11966 non-null  float64
 5   Quantity          11971 non-null  float64
 6   Total Spent       11971 non-null  float64
 7   Payment Method    12575 non-null  object 
 8   Location          12575 non-null  object 
 9   Transaction Date  12575 non-null  object 
 10  Discount Applied  8376 non-null   object 
dtypes: float64(3), object(8)
memory usage: 1.1+ MB


In [4]:
# Check missing values
df.isnull().sum()

,0
Transaction ID,0
Customer ID,0
Category,0
Item,1213
Price Per Unit,609
Quantity,604
Total Spent,604
Payment Method,0
Location,0
Transaction Date,0


**2. Handling Missing Values**

In [5]:
# Calculate missing value percentage
missing_percentage = (df.isnull().sum() / len(df)) * 100

print(missing_percentage)

Transaction ID       0.000000
Customer ID          0.000000
Category             0.000000
Item                 9.646123
Price Per Unit       4.842942
Quantity             4.803181
Total Spent          4.803181
Payment Method       0.000000
Location             0.000000
Transaction Date     0.000000
Discount Applied    33.391650
dtype: float64


In [6]:
# Handle missing values

# Categorical column
df['Item'] = df['Item'].fillna(df['Item'].mode()[0])

# Numerical columns
df['Price Per Unit'] = df['Price Per Unit'].fillna(df['Price Per Unit'].median())
df['Quantity'] = df['Quantity'].fillna(df['Quantity'].median())
df['Total Spent'] = df['Total Spent'].fillna(df['Total Spent'].median())

# Discount Applied
df['Discount Applied'] = df['Discount Applied'].fillna(
    df['Discount Applied'].mode()[0]
)

/tmp/ipykernel_2235/2428182995.py:12: FutureWarning: Downcasting object dtype arrays on .fillna, .ffill, .bfill is deprecated and will change in a future version. Call result.infer_objects(copy=False) instead. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  df['Discount Applied'] = df['Discount Applied'].fillna(


In [7]:
# Check missing values after handling
df.isnull().sum()

,0
Transaction ID,0
Customer ID,0
Category,0
Item,0
Price Per Unit,0
Quantity,0
Total Spent,0
Payment Method,0
Location,0
Transaction Date,0


**3. Duplicate Records**

In [8]:
# Check for duplicate rows
duplicates = df.duplicated().sum()

print("Number of duplicate rows:", duplicates)

Number of duplicate rows: 0


In [9]:
# Display duplicate rows
df[df.duplicated()]

,Transaction ID,Customer ID,Category,Item,Price Per Unit,Quantity,Total Spent,Payment Method,Location,Transaction Date,Discount Applied


**4. Handling Inconsistent Data**

In [10]:
# Check unique values in categorical columns

print("Category:")
print(df['Category'].unique())

print("\nPayment Method:")
print(df['Payment Method'].unique())

print("\nLocation:")
print(df['Location'].unique())

print("\nDiscount Applied:")
print(df['Discount Applied'].unique())

Category:
['Patisserie' 'Milk Products' 'Butchers' 'Beverages' 'Food' 'Furniture'
 'Electric household essentials' 'Computers and electric accessories']

Payment Method:
['Digital Wallet' 'Credit Card' 'Cash']

Location:
['Online' 'In-store']

Discount Applied:
[ True False]


In [11]:
print("Number of unique items:", df['Item'].nunique())
print(df['Item'].unique())

Number of unique items: 200
['Item_10_PAT' 'Item_17_MILK' 'Item_12_BUT' 'Item_16_BEV' 'Item_6_FOOD'
 'Item_2_BEV' 'Item_1_FOOD' 'Item_16_FUR' 'Item_22_BUT' 'Item_3_BUT'
 'Item_2_FOOD' 'Item_24_PAT' 'Item_16_MILK' 'Item_17_PAT' 'Item_13_EHE'
 'Item_7_BEV' 'Item_4_EHE' 'Item_10_FOOD' 'Item_14_FUR' 'Item_20_BUT'
 'Item_25_FUR' 'Item_14_FOOD' 'Item_22_PAT' 'Item_11_FOOD' 'Item_6_PAT'
 'Item_21_EHE' 'Item_25_BEV' 'Item_23_FOOD' 'Item_10_FUR' 'Item_11_BEV'
 'Item_23_BUT' 'Item_22_BEV' 'Item_10_EHE' 'Item_24_BUT' 'Item_8_BEV'
 'Item_3_FOOD' 'Item_12_FOOD' 'Item_16_CEA' 'Item_11_PAT' 'Item_16_BUT'
 'Item_5_CEA' 'Item_19_MILK' 'Item_23_FUR' 'Item_7_FUR' 'Item_15_CEA'
 'Item_6_MILK' 'Item_24_CEA' 'Item_22_CEA' 'Item_22_FOOD' 'Item_2_BUT'
 'Item_14_PAT' 'Item_12_PAT' 'Item_18_FOOD' 'Item_1_PAT' 'Item_4_BEV'
 'Item_22_FUR' 'Item_7_PAT' 'Item_20_CEA' 'Item_20_FOOD' 'Item_11_FUR'
 'Item_25_PAT' 'Item_7_FOOD' 'Item_21_FUR' 'Item_24_FUR' 'Item_8_MILK'
 'Item_4_FOOD' 'Item_14_BEV' 'Item_4_PAT' 'Item_4_

In [12]:
# Check unique values of categorical columns

print("Category:")
print(df['Category'].unique())

print("\nPayment Method:")
print(df['Payment Method'].unique())

print("\nLocation:")
print(df['Location'].unique())

print("\nDiscount Applied:")
print(df['Discount Applied'].unique())

Category:
['Patisserie' 'Milk Products' 'Butchers' 'Beverages' 'Food' 'Furniture'
 'Electric household essentials' 'Computers and electric accessories']

Payment Method:
['Digital Wallet' 'Credit Card' 'Cash']

Location:
['Online' 'In-store']

Discount Applied:
[ True False]


In [13]:
# Remove leading and trailing spaces from text columns

text_columns = ['Category', 'Item', 'Payment Method', 'Location']

for col in text_columns:
    df[col] = df[col].str.strip()

In [14]:
# Check categorical values after standardization

for col in text_columns:
    print(f"\n{col}:")
    print(df[col].unique())


Category:
['Patisserie' 'Milk Products' 'Butchers' 'Beverages' 'Food' 'Furniture'
 'Electric household essentials' 'Computers and electric accessories']

Item:
['Item_10_PAT' 'Item_17_MILK' 'Item_12_BUT' 'Item_16_BEV' 'Item_6_FOOD'
 'Item_2_BEV' 'Item_1_FOOD' 'Item_16_FUR' 'Item_22_BUT' 'Item_3_BUT'
 'Item_2_FOOD' 'Item_24_PAT' 'Item_16_MILK' 'Item_17_PAT' 'Item_13_EHE'
 'Item_7_BEV' 'Item_4_EHE' 'Item_10_FOOD' 'Item_14_FUR' 'Item_20_BUT'
 'Item_25_FUR' 'Item_14_FOOD' 'Item_22_PAT' 'Item_11_FOOD' 'Item_6_PAT'
 'Item_21_EHE' 'Item_25_BEV' 'Item_23_FOOD' 'Item_10_FUR' 'Item_11_BEV'
 'Item_23_BUT' 'Item_22_BEV' 'Item_10_EHE' 'Item_24_BUT' 'Item_8_BEV'
 'Item_3_FOOD' 'Item_12_FOOD' 'Item_16_CEA' 'Item_11_PAT' 'Item_16_BUT'
 'Item_5_CEA' 'Item_19_MILK' 'Item_23_FUR' 'Item_7_FUR' 'Item_15_CEA'
 'Item_6_MILK' 'Item_24_CEA' 'Item_22_CEA' 'Item_22_FOOD' 'Item_2_BUT'
 'Item_14_PAT' 'Item_12_PAT' 'Item_18_FOOD' 'Item_1_PAT' 'Item_4_BEV'
 'Item_22_FUR' 'Item_7_PAT' 'Item_20_CEA' 'Item_20_FOOD' 'I

**5. Correcting Data Types**

In [17]:
# Check current data types
df.dtypes

,0
Transaction ID,object
Customer ID,object
Category,object
Item,object
Price Per Unit,float64
Quantity,float64
Total Spent,float64
Payment Method,object
Location,object
Transaction Date,datetime64[ns]


In [16]:
# Convert Transaction Date to datetime format
df['Transaction Date'] = pd.to_datetime(df['Transaction Date'])

# Check data types again
df.dtypes

,0
Transaction ID,object
Customer ID,object
Category,object
Item,object
Price Per Unit,float64
Quantity,float64
Total Spent,float64
Payment Method,object
Location,object
Transaction Date,datetime64[ns]


**6. Outlier Detection**

In [18]:
# Check outliers using IQR method

numerical_columns = ['Price Per Unit', 'Quantity', 'Total Spent']

for col in numerical_columns:
    Q1 = df[col].quantile(0.25)
    Q3 = df[col].quantile(0.75)
    IQR = Q3 - Q1

    lower_bound = Q1 - 1.5 * IQR
    upper_bound = Q3 + 1.5 * IQR

    outliers = df[(df[col] < lower_bound) | (df[col] > upper_bound)]

    print(f"{col}:")
    print(f"Number of outliers: {len(outliers)}")
    print(f"Lower Bound: {lower_bound}")
    print(f"Upper Bound: {upper_bound}")
    print()

Price Per Unit:
Number of outliers: 0
Lower Bound: -13.0
Upper Bound: 59.0

Quantity:
Number of outliers: 0
Lower Bound: -4.5
Upper Bound: 15.5

Total Spent:
Number of outliers: 157
Lower Bound: -138.5
Upper Bound: 377.5



In [19]:
# View potential outliers in Total Spent

Q1 = df['Total Spent'].quantile(0.25)
Q3 = df['Total Spent'].quantile(0.75)
IQR = Q3 - Q1

upper_bound = Q3 + 1.5 * IQR

total_spent_outliers = df[df['Total Spent'] > upper_bound]

total_spent_outliers[['Transaction ID', 'Category', 'Item',
                      'Price Per Unit', 'Quantity', 'Total Spent']].head(20)

,Transaction ID,Category,Item,Price Per Unit,Quantity,Total Spent
27,TXN_1599706,Furniture,Item_25_FUR,41.0,10.0,410.0
120,TXN_8215469,Milk Products,Item_23_MILK,38.0,10.0,380.0
129,TXN_1112365,Food,Item_24_FOOD,39.5,10.0,395.0
133,TXN_2953434,Furniture,Item_25_FUR,41.0,10.0,410.0
135,TXN_1249742,Milk Products,Item_24_MILK,39.5,10.0,395.0
146,TXN_2565828,Food,Item_24_FOOD,39.5,10.0,395.0
339,TXN_4374445,Food,Item_25_FOOD,41.0,10.0,410.0
485,TXN_4412300,Electric household essentials,Item_23_EHE,38.0,10.0,380.0
668,TXN_5476798,Butchers,Item_2_BEV,23.0,10.0,380.0
869,TXN_1814138,Beverages,Item_25_BEV,41.0,10.0,410.0


In [20]:
print("Maximum Total Spent:", df['Total Spent'].max())

Maximum Total Spent: 410.0


**7. Normalization / Standardization**

In [21]:
from sklearn.preprocessing import MinMaxScaler

scaler = MinMaxScaler()

df[['Price_Normalized', 'Quantity_Normalized', 'Total_Spent_Normalized']] = scaler.fit_transform(
    df[['Price Per Unit', 'Quantity', 'Total Spent']]
)

df[['Price Per Unit', 'Price_Normalized',
    'Quantity', 'Quantity_Normalized',
    'Total Spent', 'Total_Spent_Normalized']].head()

,Price Per Unit,Price_Normalized,Quantity,Quantity_Normalized,Total Spent,Total_Spent_Normalized
0,18.5,0.375000,10.0,1.000000,185.0,0.444444
1,29.0,0.666667,9.0,0.888889,261.0,0.632099
2,21.5,0.458333,2.0,0.111111,43.0,0.093827
3,27.5,0.625000,9.0,0.888889,247.5,0.598765
4,12.5,0.208333,7.0,0.666667,87.5,0.203704


**8. Feature Engineering**

In [22]:
df['Year'] = df['Transaction Date'].dt.year
df['Month'] = df['Transaction Date'].dt.month
df['Day'] = df['Transaction Date'].dt.day

df[['Transaction Date', 'Year', 'Month', 'Day']].head()

,Transaction Date,Year,Month,Day
0,2024-04-08,2024,4,8
1,2023-07-23,2023,7,23
2,2022-10-05,2022,10,5
3,2022-05-07,2022,5,7
4,2022-10-02,2022,10,2


In [23]:
df['Amount_Per_Item'] = df['Total Spent'] / df['Quantity']

df[['Total Spent', 'Quantity', 'Amount_Per_Item']].head()

,Total Spent,Quantity,Amount_Per_Item
0,185.0,10.0,18.5
1,261.0,9.0,29.0
2,43.0,2.0,21.5
3,247.5,9.0,27.5
4,87.5,7.0,12.5


**9. Before vs After Data Comparison**

In [24]:
print("Dataset shape after cleaning:", df.shape)
print("\nMissing values after cleaning:")
print(df.isnull().sum())

print("\nDuplicate rows after cleaning:")
print(df.duplicated().sum())

print("\nData types after cleaning:")
print(df.dtypes)

Dataset shape after cleaning: (12575, 18)

Missing values after cleaning:
Transaction ID            0
Customer ID               0
Category                  0
Item                      0
Price Per Unit            0
Quantity                  0
Total Spent               0
Payment Method            0
Location                  0
Transaction Date          0
Discount Applied          0
Price_Normalized          0
Quantity_Normalized       0
Total_Spent_Normalized    0
Year                      0
Month                     0
Day                       0
Amount_Per_Item           0
dtype: int64

Duplicate rows after cleaning:
0

Data types after cleaning:
Transaction ID                    object
Customer ID                       object
Category                          object
Item                              object
Price Per Unit                   float64
Quantity                         float64
Total Spent                      float64
Payment Method                    object
Location           

 **10. Save Cleaned Dataset**

In [25]:
df.to_csv('retail_store_sales_cleaned.csv', index=False)

print("Cleaned dataset saved successfully!")

Cleaned dataset saved successfully!


In [26]:
import os

print("File exists:", os.path.exists('retail_store_sales_cleaned.csv'))

File exists: True


## Conclusion

The retail sales dataset was successfully cleaned and prepared for further analysis. Initially, the dataset contained 12,575 records and 11 columns with missing values in several fields. Missing values were handled appropriately, and duplicate records were checked and found to be zero.

The data types were corrected, especially the `Transaction Date` column, which was converted into a proper datetime format. Potential outliers in `Total Spent` were identified, but they were retained because they represented possible genuine high-value transactions.

Normalization was applied to the numerical variables, and new features such as `Year`, `Month`, `Day`, and `Amount_Per_Item` were created to make the dataset more useful for analysis.

After cleaning, the dataset contains 12,575 records and 18 columns with **zero missing values and zero duplicate records**. The cleaned dataset was successfully saved as `retail_store_sales_cleaned.csv`.

Overall, the data cleaning process transformed the messy retail dataset into a clean, structured, and analysis-ready dataset.
